api/service/stagedsync/stagedsync.go

```diff
@@ -0,0 +1,1321 @@
+}
+
+func (ss *StagedSync) generateStateSyncTaskQueue(bc core.BlockChain, tx kv.RwTx) error {
+	ss.stateSyncTaskQueue = queue.New(0)
+	allTasksAddedToQueue := false
+	ss.syncConfig.ForEachPeer(func(configPeer *SyncPeerConfig) (brk bool) {
+		for id, blockHash := range configPeer.blockHashes {
+			if err := ss.stateSyncTaskQueue.Put(SyncBlockTask{index: id, blockHash: blockHash}); err != nil {
+				ss.stateSyncTaskQueue = queue.New(0)
+				utils.Logger().Error().
+					Err(err).
+					Int("taskIndex", id).
+					Str("taskBlock", hex.EncodeToString(blockHash)).
+					Msg("[STAGED_SYNC] generateStateSyncTaskQueue: cannot add task")
+				break
+			}
+		}
+		// check if all block hashes added to task queue
+		if ss.stateSyncTaskQueue.Len() == int64(len(configPeer.blockHashes)) {
+			allTasksAddedToQueue = true
+			brk = true
+		}
+		return
+	})
+
+	if !allTasksAddedToQueue {
+		return ErrAddTaskFailed
+	}
+	utils.Logger().Info().
+		Int64("length", ss.stateSyncTaskQueue.Len()).
+		Msg("[STAGED_SYNC] generateStateSyncTaskQueue: finished")
+	return nil
+}
+
+// RlpDecodeBlockOrBlockWithSig decodes payload to types.Block or BlockWithSig.
+// Return the block with commitSig if set.
+func RlpDecodeBlockOrBlockWithSig(payload []byte) (*types.Block, error) {
+	var block *types.Block
+	if err := rlp.DecodeBytes(payload, &block); err == nil {
+		// received payload as *types.Block
+		return block, nil
+	}
+
+	var bws BlockWithSig
+	if err := rlp.DecodeBytes(payload, &bws); err == nil {
+		block := bws.Block
+		block.SetCurrentCommitSig(bws.CommitSigAndBitmap)
+		return block, nil
+	}
+	return nil, errors.New("failed to decode to either types.Block or BlockWithSig")
+}
+
+// CompareBlockByHash compares two block by hash, it will be used in sort the blocks
+func CompareBlockByHash(a *types.Block, b *types.Block) int {
+	ha := a.Hash()
+	hb := b.Hash()
+	return bytes.Compare(ha[:], hb[:])
+}
+
+// GetHowManyMaxConsensus will get the most common blocks and the first such blockID
+func GetHowManyMaxConsensus(blocks []*types.Block) (int, int) {
+	// As all peers are sorted by their blockHashes, all equal blockHashes should come together and consecutively.
+	curCount := 0
+	curFirstID := -1
+	maxCount := 0
+	maxFirstID := -1
+	for i := range blocks {
+		if curFirstID == -1 || CompareBlockByHash(blocks[curFirstID], blocks[i]) != 0 {
+			curCount = 1
+			curFirstID = i
+		} else {
+			curCount++
+		}
+		if curCount > maxCount {
+			maxCount = curCount
+			maxFirstID = curFirstID
+		}
+	}
+	return maxFirstID, maxCount
+}
+
+func (ss *StagedSync) getMaxConsensusBlockFromParentHash(parentHash common.Hash) *types.Block {
+	var (
+		candidateBlocks []*types.Block
+		candidateLock   sync.Mutex
+	)
+
+	ss.syncConfig.ForEachPeer(func(peerConfig *SyncPeerConfig) (brk bool) {
+		peerConfig.mux.Lock()
+		defer peerConfig.mux.Unlock()
+
+		for _, block := range peerConfig.newBlocks {
+			ph := block.ParentHash()
+			if bytes.Equal(ph[:], parentHash[:]) {
+				candidateLock.Lock()
```

Comment:
This code is not async, lock is redundant

Reply:
extra lock removed